### ISSUE TITLE
False SC2016 positive involving $(...)

### ISSUE THREAD

**Issue by scop** [koalaman/shellcheck#1726]
#### For bugs
- Rule Id (if any, e.g. SC2016): 
- My shellcheck version (`shellcheck --version` or "online"): current docker latest, 0.7.0
- [x] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/bash
foo=( $(compgen -W '$(ls /)' -- l) )
echo "${foo[@]}"

```

#### Here's what shellcheck currently says:

```
                   ^-- SC2016: Expressions don't expand in single quotes, use double quotes for that.
```

#### Here's what I wanted or expected to see:

No SC2016. This expression _does_ expand in single quotes:

```
$ bash test.bash
lib lib64
```

**Comment by Gandalf-** [comment]
You could use a directive https://github.com/koalaman/shellcheck/wiki/Directive

An alternative is to use escaping with double quotes
```shell
#!/bin/bash
foo=( $(compgen -W "\$(ls /)" -- l) )
echo "${foo[@]}"
```
This tells Shellcheck (another way) that you mean a literal `$` here. 

I think the intent behind SC2016 is good, since something like
`var='$( head -c 10 /dev/urandom )'`
and
`var="$( head -c 10 /dev/urandom )"`
could be easy for a beginner to miss.

Trade offs like this are probably necessary, but maybe we don't need this one?

**Comment by igalic** [comment]
We get something similar for code that looks like this:

```
In tasks/linux.sh line 59:
        old_version="$(apt_status '"${Version}"')"
                                  ^------------^ SC2016: Expressions don't expand in single quotes, use double quotes for that.
```

There's a lot of quoting going on here, `$()` is quoted in `""`, and `${Version}` is quoted in `""`, too.
The `''` are just for the end-user.

**Comment by koalaman** [comment]
You're right, these are all false positives. ShellCheck whitelists a few commands to try to improve the signal:noise ratio, but there's still an infinite long tail of commands where literal expansion-like arguments are correct.

Like Gandalf- says, it's most definitely intended for beginners who often assume single and double quotes are interchangeable like in Python or JavaScript. It's not at all meant to stop seasoned scripters from using single quotes when they want literal dollar signs in strings, even though that is tragically often what happens.

Unfortunately, this remains one of the most common newbie problems, it's hard for ShellCheck to avoid it (suggestions welcome).

It's obviously never great for tools to complain about correct code, but please consider helping out the newbies by instead perma-banishing it with `echo "disable=SC2016" >> ~/.shellcheckrc` when you're familiar with the difference.
